Generic coordinates

busstoptaktik opened the pull request:
This is the first in a series of  pull requests reimplementing the functionality of PR #388 "Transformation pipelines" in a more modular way.

This first step introduces the OBSERVATION data type, and a number of additions to the PJ object, intended as the basis for extended geodetic transformation functionality.

It also introduces the elemens of a new minimalistic geodetic API, orthogonal to the original proj_api.h API

QuLogic commented:
Can you remove the extra merges?

kbevers commented:
> Can you remove the extra merges?

@QuLogic Don't worry about those, they can be squashed before merging.

kbevers commented on src/proj.h:
```diff
@@ -0,0 +1,182 @@
+#endif
+
+/* Avoid preprocessor renaming and implicit type-punning: Use a union to make it explicit */
+union PJ_TRIPLET {
```
Mixed whitespace in this block

busstoptaktik replied:
Done :-)

kbevers commented on src/projects.h:
```diff
@@ -165,168 +167,320 @@ typedef struct { double u, v; } projUV;
+
+    **************************************************************************************
+
+	    For projection xxx, these are pointers to functions in the corresponding
```
Mixed whitespace in this comment section

kbevers commented on src/projects.h:
```diff
@@ -165,168 +167,320 @@ typedef struct { double u, v; } projUV;
+		a name which is currently hidden behind the magic curtain of the PROJECTION
+		macro.
+
+		As the PROJ.4 de-macroization project expands its coverage, this will change,
```
I suggest leaving this paragraph out. The odds of this message surviving the intendd de-macroization are quite high and will most likely lead to confusion for future readers of the code.

busstoptaktik replied:
Done :-)

kbevers commented on src/projects.h:
```diff
@@ -165,168 +167,320 @@ typedef struct { double u, v; } projUV;
+
+    int     datum_type;                /* PJD_UNKNOWN/3PARAM/7PARAM/GRIDSHIFT/WGS84 */
+    double  datum_params[7];           /* Parameters for 3PARAM and 7PARAM */
+    struct _pj_gi **gridlist;          /* Description needed */
```
"Description needed" is not a very useful comment. It would be good to have those descriptions made before merging. Or at least a ticket so it is not forgotten once this PR is merged.

busstoptaktik replied:
Done :-)

kbevers commented on src/projects.h:
```diff
@@ -348,35 +502,22 @@ extern struct PJ_DATUMS pj_datums[];
+
 #ifdef PJ_LIB__
-/* repetitive projection code */
+/* repetitive projection code (most of them eliminated now) */
```
Elimaniting these macros is a very good idea. Could you create a ticket that describes this effort? It will be easier to keep track of that way, and there's a bigger change someone will pick it up and do the work.

kbevers commented on src/pj_observation.c:
```diff
@@ -0,0 +1,91 @@
+
+
+/* Direction: "+1" forward, "-1" reverse, 0: roundtrip precision check (not fully done) */
+OBSERVATION pj_apply (PJ *P, int direction, OBSERVATION obs) {
```
I bet I'll have to look up the meaning og +1, -1 and 0 everytime I use it. Declaring constants like PJ_FORWARD, PJ_REVERSE and PJ_ROUNDTRIP (or something similar) will make calls to pj_apply() self-documenting.

Challenge: Come up with a self-documenting function name that isn't already use :) pj_transform would be a perfect name for this, but is already used... Anyone?

busstoptaktik replied:
Done :-)

busstoptaktik commented:
Thanks, @kbevers - I have implemented the requested improvements, and added a few missing details logically belonging to this PR.

busstoptaktik commented:
@QuLogic - does this update look sensible to you (given that commits are squashed on merge)

QuLogic commented on src/proj.h:
```diff
@@ -0,0 +1,193 @@
+#endif  /* ndef PROJECTS_H */
+
+/* Data type for generic geodetic 3D data */
+union PJ_TRIPLET;
```
Is there a reason to declare this instead of moving the `typedef` after the definition?

busstoptaktik replied:
Mostly to get a preamble style overview of "what's being defined in this file", by having declarations before the more messy definitions. During development, and the combinatorial puzzle of getting projects.h and proj.h to play nice together, the preamble sunk down below the conditional redefinitions of the proj pair/triplet types.

I have moved them back to the top, and hope you see the usefulness of the preamble style.

QuLogic commented on src/proj.h:
```diff
@@ -0,0 +1,193 @@
+OBSERVATION pj_apply (PJ *P, int direction, OBSERVATION obs);
+
+/* Direction: "+" forward, "-" reverse, 0: roundtrip precision check */
+enum pj_apply_direction {PJ_APPLY_FWD = -1, PJ_APPLY_ROUNDTRIP = 0, PJ_APPLY_INV = 1};
```
Should this not be used in `pj_apply`'s declaration (not that it makes much difference in C)? Also, I think this should go on multiple lines.

QuLogic commented on src/proj.h:
```diff
@@ -0,0 +1,193 @@
+#ifndef PROJECTS_H
+OBSERVATION pj_apply (PJ *P, int direction, OBSERVATION obs);
+
+/* Direction: "+" forward, "-" reverse, 0: roundtrip precision check */
```
Direction of what?

busstoptaktik replied:
Direction of transformation - apply the forward (1) or inverse (-1) transformation. I have added a few explanatory comments for the pj_apply function

QuLogic commented on src/projects.h:
```diff
@@ -165,168 +167,317 @@ typedef struct { double u, v; } projUV;
+    void (*spc)(LP, PJ *, struct FACTORS *);
+
+    void (*pfree)(PJ *);
+
```
I'm not sure three empty lines _plus_ the asterisks is really necessary.

QuLogic commented on src/projects.h:
```diff
@@ -165,168 +167,317 @@ typedef struct { double u, v; } projUV;
+struct ARG_list {
+    paralist *next;
+    char used;
+    char param[1];    /* This probably should be [0] to be fully standards compliant? */
```
No, this is correct; `[0]` would be an empty array. What the code does is allocate the size of this structure + the length of the string. The extra 1 here accounts for the NUL terminator on the string.

busstoptaktik replied:
Yes, I understand the reason for the 1. It just that in most historical cases of the "struct hack", param[0] was used. For many years it was a matter of heated discussion whether this was standards conforming or not. C99 (or was it C11?) solved this - but not as I seemed to remember when I wrote the comment, by requiring a 0 index, but rather by requiring the declaration to be incomplete, i.e. param[].

Hence, for C99 (11?) this is definitely non-conforming (but probably portable among all relevant platforms). For C89 it has apparently not led to any problems, despite potential non-conformality.

Given this non-straightforward state of things, I chose to remove the comment in order not to add further confusion to an already muddy matter :-)

QuLogic commented on src/projects.h:
```diff
@@ -165,168 +167,317 @@ typedef struct { double u, v; } projUV;
 #define IS_ANAL_XL_YL 01    /* derivatives of lon analytic */
 #define IS_ANAL_XP_YP 02    /* derivatives of lat analytic */
-#define IS_ANAL_HK    04    /* h and k analytic */
+#define IS_ANAL_HK  04      /* h and k analytic */
```
Not sure this needed to move?

busstoptaktik replied:
Well - its mostly a matter of putting less important stuff lower in the file. I've responded by moving a large group of related material (defines and the definition of the projCtx_t) down into the same block of the file.

QuLogic replied:
I didn't mean in the file; just the spacing left to right.

busstoptaktik replied:
Sorry - missed that one. Corrected now

busstoptaktik replied:
@QuLogic - So with this final (?) correction, can you approve the review?

busstoptaktik commented on src/Makefile.am:
```diff
@@ -72,7 +72,9 @@ libproj_la_SOURCES = \
 	jniproj.c pj_mutex.c pj_initcache.c pj_apply_vgridshift.c geodesic.c \
-	pj_strtod.c
+	pj_strtod.c \
+	\
```
Because a new block starts here: During the reimplementation of pipelines additional files will be added to this block

busstoptaktik commented on src/pj_observation.c:
```diff
@@ -0,0 +1,91 @@
+	switch (direction) {
+	    case 1:
+            obs.coo.xyz  =  pj_fwd3d (obs.coo.lpz, P);
+			return obs;
```
Yes - that's certainly an odd mix. Probably from a cut/paste operation made in hell. Cleaned up now

busstoptaktik commented on src/makefile.vc:
```diff
@@ -58,9 +58,11 @@ support = \
 	pj_strtod.obj pj_run_selftests.obj pj_generic_selftest.obj
 
+pipeline = \
+	pj_observation.obj
```
Your imagination is right :-)

busstoptaktik commented:
Thanks, @QuLogic - I have implemented (and/or elaborated on) the requested improvements, and added a small number of related corrections/improvements of consistency.

micahcochran commented on src/proj.h:
```diff
@@ -0,0 +1,202 @@
+#define TODEG(rad)  ((rad)*180/M_PI)
+#endif
+#ifndef TORAD
+#define TORAD(deg)  ((deg)*M_PI/180)
```
Would you mind using 180.0 for these two macros?  I seem to remember Visual C++ giving warnings about implicit typecasting.

busstoptaktik replied:
Thanks for spotting this @micahcochran - not being a MSVC user myself, I am probably too unaware of MSVC peculiarities.

While writing 180.0 a place, where the type promotion rules are clear, makes the code look somewhat Fortran-ish, I find it important to get rid of that kind of noise, which may divert attention away from more important warnings.

Can you approve the change?

busstoptaktik commented:
I believe this PR is ready for merging now.

rouault commented on src/proj.h:
```diff
@@ -0,0 +1,205 @@
+struct OBSERVATION {
+    PJ_SPATIOTEMPORAL coo;
+    PJ_TRIPLET anc;
+    int id;
```
Actually the same question for id and flags

busstoptaktik replied:
(and the same reply :-) )

rouault commented on src/proj.h:
```diff
@@ -0,0 +1,205 @@
+/******************************************************************************
```
I'm not sure how much we care about that but it would seem that /usr/include/proj.h is(was?) a SGI include file... https://www.google.fr/?gws_rd=ssl#q=%22%2Fusr%2Finclude%2Fproj.h%22

busstoptaktik replied:
Your google-fu is better than mine :-)

I have discussed the subject with @kbevers, and I think the common ground is that we could change the name if  anyone comes up with a good one. proj4.h was mentioned...

rouault commented on src/proj.h:
```diff
@@ -0,0 +1,205 @@
+typedef struct { double d, m,  s; }  DMS;
+
+/* Geoid undulation and deflections of the vertical */
+typedef struct { double   N,   z,  e; }  PJ_ANC_GEOIDAL;
```
What does ANC stand for here ?

busstoptaktik replied:
Ancillary data. There's a comment about it a few lines up - I have now also added some inline comments, and changed the PJ_ANC_* names for consistency

rouault commented on src/proj.h:
```diff
@@ -0,0 +1,205 @@
+#ifndef PROJECTS_H
+    OPK    opk;
+    ENH    enh;
+    PJ_ANC_GEOIDAL     nze;
```
What does nze stand for ?

busstoptaktik replied:
"N" is the commonly used designation for the geoid undulation. "z" and "e" are latinizations of the greek zeta and eta, which are commonly used designations for the components of the deflection of the vertical.

rouault commented on src/proj.h:
```diff
@@ -0,0 +1,205 @@
+#define TORAD(deg)  ((deg)*M_PI/180.0)
+#endif
+
+extern int pj_errno;                    /* global error return code */
```
Do we want to advertize this old/deprecated global way of getting errors in the new header file ?

busstoptaktik replied:
Probably a cut/paste going wrong - removed now. Thanks for spotting!

rouault commented on src/proj.h:
```diff
@@ -0,0 +1,205 @@
+
+extern int pj_errno;                    /* global error return code */
+void pj_free(PJ *P);
+PJ *pj_init_plus(const char *init_string);
```
Why not pj_init_plus_ctx ?

busstoptaktik replied:
As for pj_errno and pj_free, this is simply a cut/paste error. Eliminated now

busstoptaktik commented on src/proj.h:
```diff
@@ -0,0 +1,205 @@
+typedef struct PJconsts PJ;         /* the PJ object herself */
+
+/* Omega, Phi, Kappa: Rotations */
+typedef struct {double o, p, k;}  OPK;
```
(cf the reply to your previous comment)

busstoptaktik commented on src/proj.h:
```diff
@@ -0,0 +1,205 @@
+
+#ifndef PROJECTS_H
+/* Apply transformation to observation - in forward or inverse direction */
+OBSERVATION pj_apply (PJ *P, int direction, OBSERVATION obs);
```
@QuLogic mentioned the same thing. This is my reply to him:

I'm slightly undecided on this... I introduced the enum in response to a comment by @kbevers who felt the need for a linguistic, rather than mathematical, reminder of "what is what".

My own feeling is that having the "direction" argument directly indicate the (mathematically interpreted) power of the operator (1 forward, -1 inverse) would make ample sense. And since in geodesy numerical stability is important, I would like to have an additional version of the roundtrip test, where e.g. direction=2000 would mean "deviation from input observation for a series of 2000 roundtrips".

Hence, I would prefer to keep the direction argument an int, for future use

busstoptaktik replied:
@rouault and @QuLogic: I have now followed this suggestion by @kbevers, and think it will satisfy your comments as well :-)

busstoptaktik commented:
Hopefully ready for merging now... Thanks you to @QuLogic @kbevers @rouault and @micahcochran for excellent and useful reviews!

hobu commented:
I'm 👍 to merge this too, but it should be a squash/rebase as a single commit onto master.

kbevers commented:
Squashed and merged. Nice work, @busstoptaktik. While merging all the commits to one I've modified the initial commit message slightly: 

> * Pipeline preliminaries

> Introducing the PJ_OBSERVATION data type as the basis for generic geodetic
transformations.

> Also introducing the elements of a new minimalistic API focused on
generic geodetic transformations. This API is documented in the new
proj.h header, and is orthogonal (non-intrusive) wrt. the existing API
from proj_api.h

> Finally added a large amount of comments to the somewhat intractable
projects.h, and extended the PJ object with a number of additional
ellipsoidal parameters of general geodetic usefulness.

> The PJ elements fwdobs and invobs extend fwd3d and inv3d in a homologous
way to how fwd3d and inv3d extend fwd and inv.

I guess the next step is to document this new API. I'll write up a ticket for that right away.
